### PR TITLE
8300295: [AIX] TestDaemonDestroy fails due to !is_primordial_thread assertion

### PR DESCRIPTION
**Pull request by varada1110**
The assertion failure with exit value 134 for the test TestDaemonDestroy.java is because when the JNI_CreateJavaVM is being called by main thread on AIX it prevents the jvm creation on the primordial thread on AIX. This is solved in such a way that when JNI_CreateJavaVM is called it runs in pthread thus enabling the creation of jvm. 

JBS issue for this test : [JDK-8300295](https://bugs.openjdk.org/browse/JDK-8300295)

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8300295](https://bugs.openjdk.org/browse/JDK-8300295): [AIX] TestDaemonDestroy fails due to !is_primordial_thread assertion

### Reviewers
 * [David Holmes](https://openjdk.org/census#dholmes) (@dholmes-ora - **Reviewer**)
 * [Thomas Stuefe](https://openjdk.org/census#stuefe) (@tstuefe - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/12006/head:pull/12006` \
`$ git checkout pull/12006`

Update a local copy of the PR: \
`$ git checkout pull/12006` \
`$ git pull https://git.openjdk.org/jdk pull/12006/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 12006`

View PR using the GUI difftool: \
`$ git pr show -t 12006`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/12006.diff">https://git.openjdk.org/jdk/pull/12006.diff</a>

</details>

### CODE REVIEW

**Comment by backwaterred** [comment]
The pre-test failures on Linux and Macos do not appear related to these changes. The tests are failing with the following exception:
`compiler.lib.ir_framework.driver.irmatching.IRViolationException: There were one or multiple IR rule failures.`.

**Comment by dholmes-ora** [comment]
@varada1110  please pay attention to the bot messages and do not force-push your updates!

**Review by dholmes-ora — CHANGES_REQUESTED** [review]
It is unfortunate this is necessary.

Other tests, e.g. runtime/jni/CalleeSavedRegisters, are just disabled on AIX. If this test is to be fixed then others probably should too.

Also please address the style nits.

Thanks.

File: test/hotspot/jtreg/runtime/jni/daemonDestroy/exedaemonDestroy.c [hunk 1]
```diff
@@ -102,4 +113,24 @@ int main(int argc, char *argv[]) {
...
+int main(int argc, char *argv[]) {
+   args_list args;
+   args.argc = argc;
+   args.argv = argv;
+#ifdef AIX
+   size_t adjusted_stack_size = 1024*1024;
+   pthread_t id;
+   pthread_attr_t attr;
+   pthread_attr_init(&attr);
+   pthread_attr_setdetachstate(&attr, PTHREAD_CREATE_JOINABLE);
```
Comment (tstuefe) [inline test/hotspot/jtreg/runtime/jni/daemonDestroy/exedaemonDestroy.c:128]: Not necessary; PTHREAD_CREATE_JOINABLE is the default.

File: test/hotspot/jtreg/runtime/jni/daemonDestroy/exedaemonDestroy.c [hunk 2]
```diff
@@ -102,4 +113,24 @@ int main(int argc, char *argv[]) {
...
+   pthread_attr_setguardsize(&attr, 0);
```
Comment (tstuefe) [inline test/hotspot/jtreg/runtime/jni/daemonDestroy/exedaemonDestroy.c:129]: Why?

File: test/hotspot/jtreg/runtime/jni/daemonDestroy/exedaemonDestroy.c [hunk 3]
```diff
@@ -102,4 +113,24 @@ int main(int argc, char *argv[]) {
...
   puts("Test: DestroyJavaVM returned");
+  return 0;
+}
+
+int main(int argc, char *argv[]) {
+   args_list args;
+   args.argc = argc;
+   args.argv = argv;
+#ifdef AIX
+   size_t adjusted_stack_size = 1024*1024;
```
Comment (tstuefe) [inline test/hotspot/jtreg/runtime/jni/daemonDestroy/exedaemonDestroy.c:124]: Why "adjusted"?

File: test/hotspot/jtreg/runtime/jni/daemonDestroy/exedaemonDestroy.c [hunk 4]
```diff
@@ -102,4 +113,24 @@ int main(int argc, char *argv[]) {
...
+   pthread_t id;
+   pthread_attr_t attr;
+   pthread_attr_init(&attr);
+   pthread_attr_setdetachstate(&attr, PTHREAD_CREATE_JOINABLE);
+   pthread_attr_setguardsize(&attr, 0);
+   pthread_attr_setstacksize(&attr, adjusted_stack_size);
```
Comment (tstuefe) [inline test/hotspot/jtreg/runtime/jni/daemonDestroy/exedaemonDestroy.c:130]: Do we really need this? All other platforms seem to live fine with the default stack size.
OTOH primordial thread stack default size and pthread default stack size may differ, maybe its better to be explicit.

**Review by tstuefe — COMMENTED** [review]
Small nits, otherwise fine.

File: test/hotspot/jtreg/runtime/jni/daemonDestroy/exedaemonDestroy.c (hunk 4 above)
Comment (varada1110) [reply test/hotspot/jtreg/runtime/jni/daemonDestroy/exedaemonDestroy.c:130]: When NULL is passed as argument `pthread_create(&id, NULL, run, (void *)&args)` , the test is failing with an exit value 139. To solve that i have included the functions `pthread_attr_setdetachstate` , `pthread_attr_setguardsize` and `pthread_attr_setstacksize` to set attributes for the thread. 
Commenting `pthread_attr_setdetachstate` and `pthread_attr_setguardsize` the test has passed. I will update it here.

Thank you @tstuefe

File: test/hotspot/jtreg/runtime/jni/daemonDestroy/exedaemonDestroy.c [hunk 5]
```diff
@@ -102,4 +113,24 @@ int main(int argc, char *argv[]) {
...
+   #ifdef AIX
+   size_t adjusted_stack_size = 1024*1024;
+   pthread_t id;
+   pthread_attr_t attr;
+   pthread_attr_init(&attr);
+   pthread_attr_setdetachstate(&attr, PTHREAD_CREATE_JOINABLE);
+   pthread_attr_setguardsize(&attr, 0);
+   pthread_attr_setstacksize(&attr, adjusted_stack_size);
+   pthread_create (&id,&attr,run,(void *)&args);
+   pthread_join(id,NULL);
```
[the remark this answers is not in the document]
Comment (varada1110) [reply test/hotspot/jtreg/runtime/jni/daemonDestroy/exedaemonDestroy.c]: I have fixed the indentation. Thank you @dholmes-ora

**Review by tstuefe — CHANGES_REQUESTED** [review]
Sorry, had not noticed this before: We need error handling for the pthread routines. At the very least for pthread_create, which is the most likely to fail. Printing the errno and exiting with -1 should be fine in all cases. Would save the future analyzer of failed tests a lot of time.

**Comment by varada1110** [comment]
> Sorry, had not noticed this before: We need error handling for the pthread routines. At the very least for pthread_create, which is the most likely to fail. Printing the errno and exiting with -1 should be fine in all cases. Would save the future analyzer of failed tests a lot of time.

I have made suggested changes. Thanks @tstuefe

**Review by dholmes-ora — APPROVED** [review]
This seems functionally okay.

As I said previously there are similar tests that could be enabled for AIX and use the same mechanism. If there are in the same vicinity we may be able to add a helper file to factor this out and share it.

**Review by tstuefe — APPROVED** [review]
Okay, thanks for fixing this!

**Comment by backwaterred** [comment]
> As I said previously there are similar tests that could be enabled for AIX and use the same mechanism. If there are in the same vicinity we may be able to add a helper file to factor this out and share it.

@dholmes-ora Good suggestion. I happen to know that Varada has this in the works already. Progress is tracked here (#12302).

/sponsor

**Pull request closed without merging**